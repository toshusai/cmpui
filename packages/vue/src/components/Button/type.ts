export const ButtonVariant = {
  primary: "primary",
  secondary: "secondary",
} as const;

export type ButtonVariant = (typeof ButtonVariant)[keyof typeof ButtonVariant];

export const ButtonSize = {
  S: "S",
  M: "M",
  L: "L",
} as const;

export type ButtonSize = (typeof ButtonSize)[keyof typeof ButtonSize];

export type ButtonProps = {
  variant: ButtonVariant;
  size: ButtonSize;
  disabled: boolean;
};
