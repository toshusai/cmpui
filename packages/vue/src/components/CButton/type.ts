export const CButtonVariant = {
  primary: "primary",
  secondary: "secondary",
} as const;

export type CButtonVariant =
  (typeof CButtonVariant)[keyof typeof CButtonVariant];

export const CButtonSize = {
  S: "S",
  M: "M",
  L: "L",
} as const;

export type CButtonSize = (typeof CButtonSize)[keyof typeof CButtonSize];

export type CButtonProps = {
  variant?: CButtonVariant;
  size?: CButtonSize;
  disabled?: boolean;
};
