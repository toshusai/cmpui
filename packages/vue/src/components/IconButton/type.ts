export const IconButtonSize = {
  S: "S",
  M: "M",
} as const;

export type IconButtonSize =
  (typeof IconButtonSize)[keyof typeof IconButtonSize];

export type IconButtonProps = {
  bordered?: boolean;
  disabled?: boolean;
  rounded?: boolean;
  selected?: boolean;
  size?: IconButtonSize;
};
