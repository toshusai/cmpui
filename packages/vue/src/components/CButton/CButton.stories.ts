import type { Meta, StoryObj } from "@storybook/vue3";
import CButton from "./CButton.vue";
import { h } from "vue";

const meta = {
  title: "inputs/CButton",
  tags: ["autodocs"],
  component: CButton,
} satisfies Meta<typeof CButton>;

export default meta;
type Story = StoryObj<typeof meta>;

export const Default: Story = {
  args: {
    default: "Button",
    variant: "primary",
  },
};

export const Variants: Story = {
  render: () =>
    h(() => [
      h(CButton, { variant: "primary" }, () => "Primary"),
      h(CButton, { variant: "secondary" }, () => "Secondary"),
      h(CButton, { variant: "transparent" }, () => "Transparent"),
    ]),
};

export const Sizes: Story = {
  render: () =>
    h(() => [
      h(CButton, { size: "M" }, () => "Medium"),
      h(CButton, { size: "S" }, () => "Small"),
    ]),
};

export const Disabled: Story = {
  render: () =>
    h(() => [
      h(CButton, { variant: "primary", disabled: true }, () => "Primary"),
      h(CButton, { variant: "secondary", disabled: true }, () => "Secondary"),
    ]),
};

export const Bordered: Story = {
  render: () =>
    h(() => [
      h(CButton, { variant: "primary", bordered: true }, () => "Primary"),
      h(CButton, { variant: "secondary", bordered: true }, () => "Secondary"),
    ]),
};
