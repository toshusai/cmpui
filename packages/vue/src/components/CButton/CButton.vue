<script setup lang="ts">
import { CButtonProps } from "@toshusai/cmpui-core";
import { VNode } from "vue";

withDefaults(
  defineProps<
    CButtonProps & {
      is?: string | VNode;
    }
  >(),
  {
    is: "button",
    bordered: undefined,
    disabled: undefined,
    size: undefined,
    variant: undefined,
  },
);
</script>

<template>
  <component
    :is="is"
    class="cmpui_button"
    :data-bordered="bordered"
    :data-size="size"
    :data-variant="variant"
    :disabled="disabled"
    v-bind="$attrs"
  >
    <slot></slot>
  </component>
</template>
