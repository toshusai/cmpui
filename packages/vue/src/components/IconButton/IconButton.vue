<script setup lang="ts">
import { IconButtonProps } from "./type";

withDefaults(defineProps<IconButtonProps>(), {
  size: "M",
});
</script>

<template>
  <button
    class="cmpui_icon-button__root"
    :aria-selected="selected"
    :data-size="size"
    :disabled="disabled"
    :data-rounded="rounded"
    :data-border="border"
  >
    <slot></slot>
  </button>
</template>
