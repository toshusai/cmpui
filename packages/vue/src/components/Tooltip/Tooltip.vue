<script setup lang="ts">
import {
  ComponentPublicInstance,
  computed,
  CSSProperties,
  inject,
  onMounted,
  ref,
  useSlots,
  watch,
} from "vue";
import { tooltip } from "../../lib/tooltip";
import { useId } from "../../lib/useId";
import { isValidVNode, TOAST_PROVIDER_KEY, TooltipProvide } from "./lib";

const props = withDefaults(
  defineProps<{
    text: string;
    forceShow?: boolean;
    delay?: number;
  }>(),
  {
    forceShow: false,
    delay: 300,
  },
);

const style = ref<CSSProperties>({
  display: "none",
  position: "fixed",
  top: "0",
  left: "0",
});

const provider = inject<TooltipProvide | undefined>(
  TOAST_PROVIDER_KEY,
  undefined,
);

const slots = useSlots();

onMounted(() => {
  const children = slots.default?.().filter(isValidVNode) ?? [];
  if (children.length === 0) return;
  if (children.length > 1) {
    console.error("Tooltip can only have one child");
    return;
  }
});

const slotRef = ref<HTMLElement | ComponentPublicInstance | null>(null);
const tooltipRef = ref<HTMLElement | null>(null);
const arrowRef = ref<HTMLElement | null>(null);

const show = ref(false);

const slotElement = computed(() => {
  let el = slotRef.value;
  if (!el) return;
  if (!(el instanceof HTMLElement)) {
    el = el.$el;
    if (!(el instanceof HTMLElement)) return;
  }
  return el;
});

let i: ReturnType<typeof setTimeout> | undefined;
const handleMouseEnter = () => {
  if (provider?.show.value === true) {
    show.value = true;
    return;
  }
  i = setTimeout(() => {
    show.value = true;
  }, props.delay);
};

const handleMouseLeave = () => {
  clearTimeout(i);
  show.value = false;
};

watch(
  () => show.value,
  (show) => {
    if (provider) {
      provider.setShow(show);
    }
  },
);

watch(
  () => [slotRef.value, tooltipRef.value, arrowRef.value],
  () => {
    if (!slotRef.value) return;
    if (!tooltipRef.value) return;
    if (!slotElement.value) return;
    if (!arrowRef.value) return;
    const { cleanUp, open } = tooltip(
      tooltipRef.value,
      slotElement.value,
      arrowRef.value,
      () => {
        show.value = false;
      },
      (pos) => {
        show.value = true;

        if (pos.left !== undefined && pos.top !== undefined) {
          style.value.transform = `translate(${pos.left}px, ${pos.top}px)`;
        }
        if (pos.maxWidth !== undefined) {
          style.value.maxWidth = `${pos.maxWidth}px`;
        }
        if (pos.maxHeight !== undefined) {
          style.value.maxHeight = `${pos.maxHeight}px`;
        }
        style.value.display = "";

        if (!arrowRef.value) return;

        arrowRef.value.style.left =
          pos.arrow?.x != null ? `${pos.arrow.x}px` : "";
        arrowRef.value.style.top =
          pos.arrow?.y != null ? `${-pos.arrow.y}px` : "";

        if (pos.placement === "bottom") {
          arrowRef.value.style.transform = "rotate(180deg)";
          arrowRef.value.style.transformOrigin = "top";
          arrowRef.value.style.top = "0";
          arrowRef.value.style.bottom = "";
        } else if (pos.placement === "top") {
          arrowRef.value.style.transform = "rotate(0deg)";
          arrowRef.value.style.transformOrigin = "bottom";
          arrowRef.value.style.top = "";
          arrowRef.value.style.bottom = "-5px";
        }
      },
    );

    let stop = () => {};
    const handleOpenChange = (show: boolean) => {
      if (show) {
        open();
      } else {
        cleanUp();
        style.value.display = "none";
      }
      stop();
    };

    handleOpenChange(show.value || props.forceShow);
    stop = watch(
      () => [show.value, props.forceShow],
      ([show, forceShow]) => handleOpenChange(show || forceShow),
    );
  },
);

watch(
  () => props.forceShow,
  (forceShow) => {
    show.value = forceShow;
  },
);

const id = useId();
</script>

<template>
  <component
    :is="slots.default?.()[0]"
    ref="slotRef"
    :aria-describedby="show || forceShow ? id : undefined"
    @pointerenter="handleMouseEnter"
    @pointerleave="handleMouseLeave"
  />
  <div
    v-if="show || forceShow"
    :id="id"
    ref="tooltipRef"
    class="cmpui_tooltip__root"
    role="tooltip"
    v-bind="$attrs"
    :style="style"
  >
    {{ text }}
    <svg
      ref="arrowRef"
      class="cmpui_tooltip__arrow"
      style="position: absolute"
      width="10"
      height="5"
      viewBox="0 0 30 10"
      preserveAspectRatio="none"
    >
      <polygon points="0,0 30,0 15,10" />
    </svg>
  </div>
</template>
