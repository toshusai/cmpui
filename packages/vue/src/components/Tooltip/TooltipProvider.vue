<script setup lang="ts">
import { provide, ref } from "vue";
import { TOAST_PROVIDER_KEY } from "./lib";

const props = withDefaults(
  defineProps<{
    interval?: number;
  }>(),
  {
    interval: 500,
  },
);

const show = ref(false);

let i = 0;

const setShow = (v: boolean) => {
  if (v) {
    show.value = v;
    clearTimeout(i);
  } else {
    i = setTimeout(() => {
      show.value = v;
    }, props.interval);
  }
};

export type TooltipProvide = {
  show: typeof show;
  setShow: typeof setShow;
};

provide(TOAST_PROVIDER_KEY, {
  show: show,
  setShow: setShow,
});
</script>

<template>
  <slot />
</template>
