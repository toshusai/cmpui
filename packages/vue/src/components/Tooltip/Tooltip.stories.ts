import type { Meta, StoryObj } from "@storybook/vue3";
import IconButton from "../IconButton/IconButton.vue";
import CButton from "../CButton/CButton.vue";
import TooltipProvider from "./TooltipProvider.vue";
import Tooltip from "./Tooltip.vue";

const meta = {
  component: Tooltip,
  args: {
    text: "tooltip",
  },
} satisfies Meta<typeof Tooltip>;

export default meta;
type Story = StoryObj<typeof meta>;

const html = String.raw;

export const Default: Story = {
  render: () => ({
    components: { CButton, Tooltip, TooltipProvider },
    template: `
        <Tooltip text="Tooltip">
          <CButton>
            Hover me
          </CButton>
        </Tooltip>
    `,
  }),
};

export const Flip: Story = {
  render: () => ({
    components: { CButton, Tooltip, TooltipProvider },
    template: `
        <Tooltip text="Tooltip">
          <CButton style="position: absolute; bottom: 4px; transform: translateX(-50%);">
            Hover me
          </CButton>
        </Tooltip>
    `,
  }),
};

export const AutoPlacement: Story = {
  render: () => ({
    components: { CButton, Tooltip, TooltipProvider },
    template: `
        <Tooltip text="Tooltip can be placed anywhere">
          <CButton style="position: absolute; bottom: 4px; right: 4px">
            Hover me
          </CButton>
        </Tooltip>
    `,
  }),
};

export const MultipleItems: Story = {
  render: () => ({
    components: { IconButton, Tooltip, TooltipProvider },
    template: html`
      <TooltipProvider>
        <div class="cmpui_icon-button__group">
          <Tooltip text="pointer">
            <IconButton>
              <i class="ti ti-pointer"></i>
            </IconButton>
          </Tooltip>
          <Tooltip text="typography">
            <IconButton>
              <i class="ti ti-typography"></i>
            </IconButton>
          </Tooltip>
          <Tooltip text="italic">
            <IconButton>
              <i class="ti ti-italic"></i>
            </IconButton>
          </Tooltip>
        </div>
      </TooltipProvider>
    `,
  }),
};
