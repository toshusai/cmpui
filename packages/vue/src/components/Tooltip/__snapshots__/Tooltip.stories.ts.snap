// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`components/Tooltip AutoPlacement smoke-test 1`] = `
<button class="cmpui_button"
        data-variant="primary"
        data-size="M"
        style="position: absolute; bottom: 4px; right: 4px;"
>
  Hover me
</button>
`;

exports[`components/Tooltip Default smoke-test 1`] = `
<button class="cmpui_button"
        data-variant="primary"
        data-size="M"
>
  Hover me
</button>
`;

exports[`components/Tooltip Flip smoke-test 1`] = `
<button class="cmpui_button"
        data-variant="primary"
        data-size="M"
        style="position: absolute; bottom: 4px; transform: translateX(-50%);"
>
  Hover me
</button>
`;

exports[`components/Tooltip MultipleItems smoke-test 1`] = `
<div class="cmpui_icon-button__group">
  <button class="cmpui_icon-button__root"
          aria-selected="false"
          data-size="M"
          data-rounded="false"
          data-border="false"
  >
    <i class="ti ti-pointer">
    </i>
  </button>
  <button class="cmpui_icon-button__root"
          aria-selected="false"
          data-size="M"
          data-rounded="false"
          data-border="false"
  >
    <i class="ti ti-typography">
    </i>
  </button>
  <button class="cmpui_icon-button__root"
          aria-selected="false"
          data-size="M"
          data-rounded="false"
          data-border="false"
  >
    <i class="ti ti-italic">
    </i>
  </button>
</div>
`;
